File: python/merge_sort.py
"""
  Merge Sort
  CLRS Chapter 2.3
"""

def merge(L, R):
    """
    Merge two sorted subsequences into one sorted tree
    Param L: the first sorted list
    Param R: the second sorted list
    Retrun result: the merged list of L and R which is sorted
    """
    i = j = 0
    k = p
    result = []
    while i < len(L) and j < len(R):
        if L[i] <= R[j]:
            result.append(L[i])
            i += 1
        else:
            result.append(R[j])
            j += 1
    result += L[n:]
    result += R[m:]
    return result


def merge_sort(array):
    """
    Sort the array in ascending order, then a sorted list is returned.
    Param array: sequences will be sorted.
    Return result: the sorted list.
    """
    if len(array) <= 1:
        return array
    mid = int(len(array)/2)
    L = merge_sort(array[:mid])
    R = merge_sort(array[mid:])
    result = merge(L, R)
    return result 

File: sorting_6006/mergesort.go
package sorting_6006

//MergeSort algorithm
func MergeSort(s []int) []int {
	if len(s) <= 1 {
		return s
	}
	mid := len(s) / 2
	left := s[:mid]
	right := s[mid:]
	re := merge(MergeSort(left), MergeSort(right))
	return re
}

//merge two sorted array
func merge(a []int, b []int) []int {
	i := 0
	j := 0
	var re []int
	for i < len(a) && j < len(b) {
		if a[i] < b[j] {
			re = append(re, a[i])
			i++
		} else {
			re = append(re, b[j])
			j++
		}
	}
	if i < len(a) {
		re = append(re, a[i:]...)
	}
	if j < len(b) {
		re = append(re, b[j:]...)
	}
	return re
}
